nlp/transform.py

```
@@ -0,0 +1,163 @@
+	z - set to 0
+	depth - set to 0
+
+		(x,y+height).   (x+width, y+height)
+		---------------
+		|			  |
+		|			  |
+		|			  |
+		|			  |		
+		|			  |
+		________________
+		(x,y)			(x+width,y)
+
+	these test cases are for a 100*100 image but generic code is built to run any height and width
+	>>> normalize_x_and_y(image,100,100,50,50)
+	1.0 0.0 0.0 0.0 0.0 0.0
+	>>> normalize_x_and_y(image,10,90,10,90)
+	-0.8 -0.8 1.6 1.6 0.0 0.0
+	>>> normalize_x_and_y(image,90,10,10,90)
+	xmin is greater than xmax
+	>>> normalize_x_and_y(image,110,190,10,40)
+	xmin is greater than image width
+	>>> normalize_x_and_y(image,-10,10,10,90)
+	xmin < 0
+	>>> normalize_x_and_y(image,10,90,90,10)
+	ymin is greater than ymax
+	>>> normalize_x_and_y(image,101,150,50,100)
+	xmin is greater than image width
+	>>> normalize_x_and_y(image,10,90,10,190)
+	ymax is greater than image height
+	>>> normalize_x_and_y(iamge,0,200,100,100)
+	xmax is greater than image width
+	>>> normalize_x_and_y(image,0,100,-10,100)
+	ymin < 0
+	>>> normalize_x_and_y(image,50,50,50,50)
+	0.0 0.0 0.0 0.0 0.0 0.0
+	>>> normalize_x_and_y(image,50,100,50,100)
+	0.0 0.0 1.0 1.0 0.0 0.0
+	>>> normalize_x_and_y(image,0,100,0,100)
+	-1.0 -1.0 2.0 2.0 0.0 0.0
+	
+	"""
+
+	#TODO: Check if I really need this next line or not
+	im = Image.open(image)
+	im_width, im_height = im.size
+	if (xmin > xmax):
+		print ("xmin is greater than xmax")
+		return ("error")
+
+	if (ymin>ymax):
+		print("ymin is greater than ymax")
+		return ("error")
+
+	if (xmin > im_width):
+		print("xmin is greater than image width")
+		return ("error")
+
+	if (xmax > im_width):
+		print("xmax is greater than image width")
+		return ("error")
+
+	if (ymin > im_height):
+		print ("ymin is greater than image height")
+		return ("error")
+	if (ymax > im_height):
+		print ("ymax is greater than image height")
+		return ("error")
+
+	if (xmin < 0):
+		print ("xmin < 0")
+		return ("error")
+
+	if (ymin < 0):
+		print ("ymin < 0")
+		return ("error")
```

Instead of returning an error, these lines should be replaced by `assert` statements or throw an error. This is better from the caller's perspective, because they can handle bad input with a try catch instead of parsing the tuple/working with strings.

---

It's a bad practice to pass around errors as data, better to use exceptions (baring performance concerns)